File: android/src/main/java/com/zhuge/reactnativezhuge/RNZhugeio.java
package com.zhuge.reactnativezhuge;

import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.zhuge.analysis.stat.ZhugeSDK;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Set;

/**
 * 将Zhuge SDK 方法暴露至React native
 * Created by Jiaokang on 2018/2/27.
 */

public class RNZhugeio extends ReactContextBaseJavaModule {

    private static final String TAG = "RNZhugeio";
    private static final String MODULE_NAME = "Zhugeio";

    public RNZhugeio(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @ReactMethod
    public void init(String appkey, String channel) {
        ZhugeParam param = new ZhugeParam.Builder()
        .appKey(appkey)
        .appChannel(channel)
        .build();
        ZhugeSDK.getInstance().initWithParam(getReactApplicationContext(),param);
    }

    @ReactMethod
    public void openLog() {
        ZhugeSDK.getInstance().openLog();
    }

    @ReactMethod
    public void openDebug() {
        ZhugeSDK.getInstance().openDebug();
    }

    @ReactMethod
    public void setUploadURL(String url, String backupUrl) {
        ZhugeSDK.getInstance().setUploadURL(url, backupUrl);
    }

    @ReactMethod
    public void identify(String uid, ReadableMap pro) {
        debug("identify " + uid);
        HashMap<String, Object> properties = null;
        if (pro != null) {
            properties = pro.toHashMap();
        } else {
            properties = new HashMap<>();
        }
        ZhugeSDK.getInstance().identify(null, uid, properties);
    }

    @ReactMethod
    public void track(String name, ReadableMap pro) {
        debug("track " + name);
        HashMap<String, Object> properties = null;
        if (pro != null) {
            properties = pro.toHashMap();
        } else {
            properties = new HashMap<>();
        }
        ZhugeSDK.getInstance().track(null, name, properties);
    }

    @ReactMethod
    public void trackRevenue(ReadableMap pro) {
        debug("track revenue");
        HashMap<String, Object> properties = null;
        if (pro != null) {
            properties = pro.toHashMap();
        } else {
            properties = new HashMap<>();
        }
        ZhugeSDK.getInstance().trackRevenue(null, properties);
    }

    @ReactMethod
    public void startTrack(String name) {
        debug("startTrack " + name);
        ZhugeSDK.getInstance().startTrack(name);
    }

    @ReactMethod
    public void endTrack(String name, ReadableMap pro) {
        debug("endTrack : " + name);
        JSONObject properties = null;
        if (pro != null) {
            HashMap<String, Object> map = pro.toHashMap();
            Set<String> keySet = map.keySet();
            properties = new JSONObject();
            try {
                for (String key : keySet) {
                    properties.put(key, map.get(key));
                }
            } catch (Exception e) {
                Log.e(TAG, "convert map to json error", e);
            }
        }
        ZhugeSDK.getInstance().endTrack(name, properties);
    }

    @ReactMethod
    public void setUtm(ReadableMap utm) {
        JSONObject info = new JSONObject();
        if (utm != null) {
            HashMap<String, Object> map = utm.toHashMap();
            Set<String> keySet = map.keySet();
            try {
                for (String key : keySet) {
                    info.put(key, map.get(key));
                }
            } catch (Exception e) {
                Log.e(TAG, "convert map to json error", e);
            }
        }
        ZhugeSDK.getInstance().setUtm(info);
    }

    @Override
    public String getName() {
        return MODULE_NAME;
    }

    private void debug(String info) {
        Log.d(TAG, info);
    }
}

File: ios/RNZhugeio.m
#import "RNZhugeio.h"

#if __has_include("Zhuge.h")
#import "Zhuge.h"
#else
#import <ZhugeioAnalytics/Zhuge.h>
#endif

@implementation RNZhugeio

RCT_EXPORT_MODULE(Zhugeio)

RCT_EXPORT_METHOD(openDebug){
    NSLog(@"openDebug");
    [[Zhuge sharedInstance].config setDebug:true];
}

RCT_EXPORT_METHOD(openLog){
    [Zhuge sharedInstance].config.enableLoger = YES;
}

RCT_EXPORT_METHOD(init:(NSString *)appKey channel:(NSString *)channel){
  [[Zhuge sharedInstance] startWithAppKey:appKey launchOptions:nil];
}

RCT_EXPORT_METHOD(setUploadURL:(NSString *)url backupURL:(NSString *)backup){
    NSLog(@"setUploadURL with %@ and %@",url,backup);
    [[Zhuge sharedInstance]setUploadURL:url andBackupUrl:backup];
}

RCT_EXPORT_METHOD(identify:(NSString *)uid properties:(NSDictionary *)pro){
    NSLog(@"identify with %@",uid);
    [[Zhuge sharedInstance]identify:uid properties:pro];
}

RCT_EXPORT_METHOD(trackRevenue:(NSDictionary *)pro){
    NSLog(@"itrack with revenue");
    [[Zhuge sharedInstance] trackRevenue:pro];
}


RCT_EXPORT_METHOD(track:(NSString *)name properties:(NSDictionary *)pro)
{
    Zhuge *zhuge = [Zhuge sharedInstance];
    NSLog(@"track with %@",name);
    [zhuge track:name properties:pro];
}

RCT_EXPORT_METHOD(setUtm:(NSDictionary *)utm){
    Zhuge *zhuge = [Zhuge sharedInstance];
    NSLog(@"setUtm with %@",utm);
//    [zhuge setUtm:utm];
}

RCT_EXPORT_METHOD(startTrack:(NSString *)name){
    NSLog(@"startTrack with %@",name);
    [[Zhuge sharedInstance]startTrack:name];
}

RCT_EXPORT_METHOD(endTrack:(NSString *)name properties:(NSDictionary *)pro){
    NSLog(@"endTrack with %@",name);
    [[Zhuge sharedInstance] endTrack:name properties:pro];
}

@end
